src/gc/collector.cpp

```diff
@@ -342,26 +416,121 @@ void markPhase() {
 #endif
 }
 
+void finalizationOrderingFirstPass(Box* obj) {
+    std::shared_ptr<FirstPhaseStack> stack = std::make_shared<FirstPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        visitByGCKind(p, visitor);
+    }
+}
+
+void finalizationOrderingSecondPass(Box* obj) {
+    std::shared_ptr<SecondPhaseStack> stack = std::make_shared<SecondPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        GCAllocation* al = GCAllocation::fromUserData(p);
+        assert(orderingState(al) != FinalizationState::UNREACHABLE);
+        visitByGCKind(p, visitor);
+    }
+}
+
+// Implementation of PyPy's finalization ordering algorithm:
+// http://pypy.readthedocs.org/en/latest/discussion/finalizer-order.html
+void finalizationOrderingPhase() {
+    // TODO: Replace with iterator or something that takes a lambda for memory efficiency?.
+    std::vector<Box*> objects_with_finalizers;
+    global_heap.getOrderedFinalizers(objects_with_finalizers);
```

Comment:
I don't think so, because unreachable objects with finalizers can mark other unreachable objects that it references. I think we'd need to consider other objects like tracking finalizable objects as they get created.